core/src/main/java/com/datastrato/gravitino/storage/relational/service/SchemaMetaService.java

```diff
@@ -0,0 +1,283 @@
+  }
+
+  public void insertSchema(SchemaEntity schemaEntity, boolean overwrite) {
+    try {
+      NameIdentifier.checkSchema(schemaEntity.nameIdentifier());
+      Long metalakeId =
+          SessionUtils.getWithoutCommit(
+              MetalakeMetaMapper.class,
+              mapper -> mapper.selectMetalakeIdMetaByName(schemaEntity.namespace().level(0)));
+      if (metalakeId == null) {
+        throw new NoSuchEntityException(
+            NoSuchEntityException.NO_SUCH_ENTITY_MESSAGE,
+            Entity.EntityType.METALAKE.name().toLowerCase(),
+            schemaEntity.namespace().level(0));
+      }
+      Long catalogId =
+          SessionUtils.getWithoutCommit(
+              CatalogMetaMapper.class,
+              mapper ->
+                  mapper.selectCatalogIdByMetalakeIdAndName(
+                      metalakeId, schemaEntity.namespace().level(1)));
+      if (catalogId == null) {
+        throw new NoSuchEntityException(
+            NoSuchEntityException.NO_SUCH_ENTITY_MESSAGE,
+            Entity.EntityType.CATALOG.name().toLowerCase(),
+            schemaEntity.namespace());
+      }
+
+      SessionUtils.doWithCommit(
+          SchemaMetaMapper.class,
+          mapper -> {
+            SchemaPO po =
+                POConverters.initializeSchemaPOWithVersion(schemaEntity, metalakeId, catalogId);
+            if (overwrite) {
+              mapper.insertSchemaMetaOnDuplicateKeyUpdate(po);
+            } else {
+              mapper.insertSchemaMeta(po);
+            }
+          });
+    } catch (RuntimeException re) {
+      if (re.getCause() != null
+          && re.getCause().getCause() != null
+          && re.getCause().getCause() instanceof SQLIntegrityConstraintViolationException) {
+        // TODO We should make more fine-grained exception judgments
+        // Usually throwing `SQLIntegrityConstraintViolationException` means that
+        // SQL violates the constraints of `primary key` and `unique key`.
+        // We simply think that the entity already exists at this time.
+        throw new EntityAlreadyExistsException(
+            String.format("Schema entity: %s already exists", schemaEntity.nameIdentifier()));
+      }
+      throw re;
+    }
+  }
+
+  public <E extends Entity & HasIdentifier> SchemaEntity updateSchema(
+      NameIdentifier identifier, Function<E, E> updater) throws IOException {
+    NameIdentifier.checkSchema(identifier);
+    String metalakeName = identifier.namespace().level(0);
+    String catalogName = identifier.namespace().level(1);
+    String schemaName = identifier.name();
+
+    Long metalakeId =
+        SessionUtils.getWithoutCommit(
+            MetalakeMetaMapper.class, mapper -> mapper.selectMetalakeIdMetaByName(metalakeName));
+    if (metalakeId == null) {
+      throw new NoSuchEntityException(
+          NoSuchEntityException.NO_SUCH_ENTITY_MESSAGE,
+          Entity.EntityType.METALAKE.name().toLowerCase(),
+          metalakeName);
+    }
+
+    Long catalogId =
```

Comment:
Could you extract a method to reduce the duplicated code?
```
Long catalogId =
Comment
Could you extract a method to reduce the duplicated code?
```

```
 
        SessionUtils.getWithoutCommit(
            CatalogMetaMapper.class,
            mapper -> mapper.selectCatalogIdByMetalakeIdAndName(metalakeId, catalogName));
    if (catalogId == null) {
      throw new NoSuchEntityException(
          NoSuchEntityException.NO_SUCH_ENTITY_MESSAGE,
          Entity.EntityType.CATALOG.name().toLowerCase(),
          identifier.namespace().toString());
    }
```

Reply:
Let me try

Reply:
@qqqttt123 Done